CSV import bug for 3.3.2+

I used NEO4J 3.3.1 on my mac to use neo4j-admin import shell to built graphs from CSV files and everything works fine. (followed by https://neo4j.com/docs/operations-manual/current/tools/import/)

I wanted to build the same database on my Ubuntu box(14.04) with NEO4J 3.3.2 so I scp all the CSV files from mac to ubuntu box and used the same shell to import CSV.

However, the importing data failed and neo4j complained the below ERRORS.
I tried to upgrade to 3.3.3 but it still failed in the same way. 
Finally I downgraded to 3.3.1 like what I have in Mac and it magically worked.


> WARNING Import failed. The store files in /home/mlo/neo4j-community-3.3.3/data/databases/tw_ml_v3.db are left as they are, although they are likely in an unusable state. Starting a database on these store files will likely fail or observe inconsistent records so start at your own risk or delete the store manually
> org.neo4j.unsafe.impl.batchimport.input.InputException: ERROR in input
>   data source: BufferedCharSeeker[source:/home/mlo/tw_ml_v3/merch_user_table.csv, position:2699377, line:9385]
>   in field: photo_id_number:string:15
>   for header: [id:ID(Merch_user), username:string, email:string, individual_name:string, store_url:string, is_banned:string, display_name:string, product_categories:string, date_last_active:string, tax_certificate:string, company_name:string, business_license:string, photo_id:string, warehouse_locations:string, photo_id_number:string, bd_rep:string]
>   raw field value: ??
>   original error: String index out of range: -283
> 	at org.neo4j.unsafe.impl.batchimport.input.csv.InputEntityDeserializer.deserializeNextFromSource(InputEntityDeserializer.java:178)
> 	at org.neo4j.unsafe.impl.batchimport.input.csv.InputEntityDeserializer.fetchNextOrNull(InputEntityDeserializer.java:77)
> 	at org.neo4j.unsafe.impl.batchimport.input.csv.InputEntityDeserializer.fetchNextOrNull(InputEntityDeserializer.java:41)
> 	at org.neo4j.helpers.collection.PrefetchingIterator.peek(PrefetchingIterator.java:60)
> 	at org.neo4j.helpers.collection.PrefetchingIterator.hasNext(PrefetchingIterator.java:46)
> 	at org.neo4j.unsafe.impl.batchimport.input.csv.ParallelInputEntityDeserializer.lambda$new$0(ParallelInputEntityDeserializer.java:106)
> 	at org.neo4j.unsafe.impl.batchimport.staging.TicketedProcessing.lambda$submit$1(TicketedProcessing.java:103)
> 	at org.neo4j.unsafe.impl.batchimport.executor.DynamicTaskExecutor$Processor.run(DynamicTaskExecutor.java:237)
> 	Suppressed: org.neo4j.unsafe.impl.batchimport.executor.TaskExecutionPanicException: Executor has been shut down in panic
> 		at org.neo4j.unsafe.impl.batchimport.executor.DynamicTaskExecutor.assertHealthy(DynamicTaskExecutor.java:146)
> 		at org.neo4j.unsafe.impl.batchimport.executor.DynamicTaskExecutor.submit(DynamicTaskExecutor.java:126)
> 		at org.neo4j.unsafe.impl.batchimport.staging.ProcessorStep.receive(ProcessorStep.java:78)
> 		at org.neo4j.unsafe.impl.batchimport.staging.Downstream.send(Downstream.java:56)
> 		at org.neo4j.unsafe.impl.batchimport.staging.SendDownstream.apply(SendDownstream.java:58)
> 		at org.neo4j.unsafe.impl.batchimport.staging.SendDownstream.apply(SendDownstream.java:26)
> 		at org.neo4j.concurrent.WorkSync.doSynchronizedWork(WorkSync.java:231)
> 		at org.neo4j.concurrent.WorkSync.tryDoWork(WorkSync.java:157)
> 		at org.neo4j.concurrent.WorkSync.applyAsync(WorkSync.java:122)
> 		at org.neo4j.unsafe.impl.batchimport.staging.ProcessorStep.sendDownstream(ProcessorStep.java:162)
> 		at org.neo4j.unsafe.impl.batchimport.staging.ProcessorStep.access$100(ProcessorStep.java:45)
> 		at org.neo4j.unsafe.impl.batchimport.staging.ProcessorStep$Sender.send(ProcessorStep.java:223)
> 		at org.neo4j.unsafe.impl.batchimport.PropertyEncoderStep.process(PropertyEncoderStep.java:137)
> 		at org.neo4j.unsafe.impl.batchimport.PropertyEncoderStep.process(PropertyEncoderStep.java:52)
> 		at org.neo4j.unsafe.impl.batchimport.staging.ProcessorStep.lambda$receive$1(ProcessorStep.java:85)
> 		... 1 more
> 	[CIRCULAR REFERENCE:org.neo4j.unsafe.impl.batchimport.input.InputException: ERROR in input
>   data source: BufferedCharSeeker[source:/home/mlo/tw_ml_v3/merch_user_table.csv, position:2699377, line:9385]
>   in field: photo_id_number:string:15
>   for header: [id:ID(Merch_user), username:string, email:string, individual_name:string, store_url:string, is_banned:string, display_name:string, product_categories:string, date_last_active:string, tax_certificate:string, company_name:string, business_license:string, photo_id:string, warehouse_locations:string, photo_id_number:string, bd_rep:string]
>   raw field value: ??
>   original error: String index out of range: -283]
> 	[CIRCULAR REFERENCE:org.neo4j.unsafe.impl.batchimport.executor.TaskExecutionPanicException: Executor has been shut down in panic]
> 	Suppressed: org.neo4j.unsafe.impl.batchimport.executor.TaskExecutionPanicException: Executor has been shut down in panic
> 		at org.neo4j.unsafe.impl.batchimport.executor.DynamicTaskExecutor.assertHealthy(DynamicTaskExecutor.java:146)
> 		at org.neo4j.unsafe.impl.batchimport.executor.DynamicTaskExecutor.submit(DynamicTaskExecutor.java:126)
> 		at org.neo4j.unsafe.impl.batchimport.staging.ProcessorStep.receive(ProcessorStep.java:78)
> 		at org.neo4j.unsafe.impl.batchimport.staging.Downstream.send(Downstream.java:56)
> 		at org.neo4j.unsafe.impl.batchimport.staging.SendDownstream.apply(SendDownstream.java:58)
> 		at org.neo4j.unsafe.impl.batchimport.staging.SendDownstream.apply(SendDownstream.java:26)
> 		at org.neo4j.concurrent.WorkSync.doSynchronizedWork(WorkSync.java:231)
> 		at org.neo4j.concurrent.WorkSync.tryDoWork(WorkSync.java:157)
> 		at org.neo4j.concurrent.WorkSync.applyAsync(WorkSync.java:122)
> 		at org.neo4j.unsafe.impl.batchimport.staging.ProcessorStep.sendDownstream(ProcessorStep.java:162)
> 		at org.neo4j.unsafe.impl.batchimport.staging.ProcessorStep.access$100(ProcessorStep.java:45)
> 		at org.neo4j.unsafe.impl.batchimport.staging.ProcessorStep$Sender.send(ProcessorStep.java:223)
> 		at org.neo4j.unsafe.impl.batchimport.PropertyEncoderStep.process(PropertyEncoderStep.java:137)
> 		at org.neo4j.unsafe.impl.batchimport.PropertyEncoderStep.process(PropertyEncoderStep.java:52)
> 		at org.neo4j.unsafe.impl.batchimport.staging.ProcessorStep.lambda$receive$1(ProcessorStep.java:85)
> 		... 1 more
> 	[CIRCULAR REFERENCE:org.neo4j.unsafe.impl.batchimport.input.InputException: ERROR in input
>   data source: BufferedCharSeeker[source:/home/mlo/tw_ml_v3/merch_user_table.csv, position:2699377, line:9385]
>   in field: photo_id_number:string:15
>   for header: [id:ID(Merch_user), username:string, email:string, individual_name:string, store_url:string, is_banned:string, display_name:string, product_categories:string, date_last_active:string, tax_certificate:string, company_name:string, business_license:string, photo_id:string, warehouse_locations:string, photo_id_number:string, bd_rep:string]
>   raw field value: ??
>   original error: String index out of range: -283]
> 	[CIRCULAR REFERENCE:org.neo4j.unsafe.impl.batchimport.executor.TaskExecutionPanicException: Executor has been shut down in panic]
> Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: -283
> 	at java.lang.String.<init>(String.java:196)
> 	at org.neo4j.csv.reader.Extractors$StringExtractor.extract0(Extractors.java:329)
> 	at org.neo4j.csv.reader.Extractors$AbstractSingleValueExtractor.extract(Extractors.java:290)
> 	at org.neo4j.csv.reader.BufferedCharSeeker.tryExtract(BufferedCharSeeker.java:283)
> 	at org.neo4j.unsafe.impl.batchimport.input.csv.InputEntityDeserializer.deserializeNextFromSource(InputEntityDeserializer.java:138)
> 	... 7 more
> unexpected error: ERROR in input
>   data source: BufferedCharSeeker[source:/home/mlo/tw_ml_v3/merch_user_table.csv, position:2699377, line:9385]
>   in field: photo_id_number:string:15
>   for header: [id:ID(Merch_user), username:string, email:string, individual_name:string, store_url:string, is_banned:string, display_name:string, product_categories:string, date_last_active:string, tax_certificate:string, company_name:string, business_license:string, photo_id:string, warehouse_locations:string, photo_id_number:string, bd_rep:string]
>   raw field value: ??
>   original error: String index out of range: -283

--- a/community/csv/src/main/java/org/neo4j/csv/reader/BufferedCharSeeker.java
+++ b/community/csv/src/main/java/org/neo4j/csv/reader/BufferedCharSeeker.java
@@ -22,9 +22,11 @@ package org.neo4j.csv.reader;
 import java.io.BufferedReader;
 import java.io.IOException;
 import java.io.Reader;
+import java.util.Arrays;
 
 import org.neo4j.csv.reader.Source.Chunk;
 
+import static java.lang.Character.isWhitespace;
 import static java.lang.String.format;
 import static org.neo4j.csv.reader.Mark.END_OF_LINE_CHARACTER;
 
@@ -37,7 +39,6 @@ public class BufferedCharSeeker implements CharSeeker
     private static final char EOL_CHAR_2 = '\r';
     private static final char EOF_CHAR = (char) -1;
     private static final char BACK_SLASH = '\\';
-    private static final char WHITESPACE = ' ';
 
     private char[] buffer;
     private int dataLength;
@@ -99,11 +100,14 @@ public class BufferedCharSeeker implements CharSeeker
             ch = nextChar( skippedChars );
             if ( quoteDepth == 0 )
             {   // In normal mode, i.e. not within quotes
-                if ( isWhitespace( ch ) && trim )
-                {
+                if ( ch == untilChar )
+                {   // We found a delimiter, set marker and return true
+                    return setMark( mark, endOffset, skippedChars, ch, isQuoted );
+                }
+                else if ( trim && isWhitespace( ch ) )
+                {   // Only check for left+trim whitespace as long as we haven't found a non-whitespace character
                     if ( seekStartPos == bufferPos - 1/* -1 since we just advanced one */ )
-                    {
-                        // We found a whitespace, which was the first of the value and we've been told to trim that off
+                    {   // We found a whitespace, which is before the first non-whitespace of the value and we've been told to trim that off
                         seekStartPos++;
                     }
                 }
@@ -124,18 +128,12 @@ public class BufferedCharSeeker implements CharSeeker
                     }
                     break;
                 }
-                else if ( ch == untilChar )
-                {   // We found a delimiter, set marker and return true
-                    return setMark( mark, endOffset, skippedChars, ch, isQuoted );
+                else if ( isQuoted )
+                {   // This value is quoted, i.e. started with a quote and has also seen a quote
+                    throw new DataAfterQuoteException( this,
+                            new String( buffer, seekStartPos, bufferPos - seekStartPos ) );
                 }
-                else
-                {   // This is a character to include as part of the current value
-                    if ( isQuoted )
-                    {   // This value is quoted, i.e. started with a quote and has also seen a quote
-                        throw new DataAfterQuoteException( this,
-                                new String( buffer, seekStartPos, bufferPos - seekStartPos ) );
-                    }
-                }
+                // else this is a character to include as part of the current value
             }
             else
             {   // In quoted mode, i.e. within quotes
@@ -217,7 +215,18 @@ public class BufferedCharSeeker implements CharSeeker
 
     private boolean isWhitespace( int ch )
     {
-        return ch == WHITESPACE;
+        return ch == ' ' |
+                ch == Character.SPACE_SEPARATOR |
+                ch == Character.PARAGRAPH_SEPARATOR |
+                ch == '\u00A0' |
+                ch == '\u001C' |
+                ch == '\u001D' |
+                ch == '\u001E' |
+                ch == '\u001F' |
+                ch == '\u2007' |
+                ch == '\u202F' |
+                ch == '\t';
+
     }
 
     private void repositionChar( int offset, int stepsBack )

--- a/community/csv/src/main/java/org/neo4j/csv/reader/BufferedCharSeeker.java
+++ b/community/csv/src/main/java/org/neo4j/csv/reader/BufferedCharSeeker.java
@@ -194,15 +194,21 @@ public class BufferedCharSeeker implements CharSeeker
 
     private boolean setMark( Mark mark, int endOffset, int skippedChars, int ch, boolean isQuoted )
     {
-        int pos = (trim ? rtrim( bufferPos ) : bufferPos) - endOffset - skippedChars;
+        int pos = (trim ? rtrim() : bufferPos) - endOffset - skippedChars;
         mark.set( seekStartPos, pos, ch, isQuoted );
         return true;
     }
 
-    private int rtrim( int start )
+    /**
+     * Starting from the current position, {@link #bufferPos}, scan backwards as long as whitespace is found.
+     * Although it cannot scan further back than the start of this field is, i.e. {@link #seekStartPos}.
+     *
+     * @return the right index of the value to pass into {@link Mark}. This is only called if {@link Configuration#trimStrings()} is {@code true}.
+     */
+    private int rtrim()
     {
-        int index = start;
-        while ( isWhitespace( buffer[index - 1 /*bufferPos has advanced*/ - 1 /*don't check the last read char (delim or EOF)*/] ) )
+        int index = bufferPos;
+        while ( index - 1 > seekStartPos && isWhitespace( buffer[index - 1 /*bufferPos has advanced*/ - 1 /*don't check the last read char (delim or EOF)*/] ) )
         {
             index--;
         }

--- a/community/csv/src/test/java/org/neo4j/csv/reader/BufferedCharSeekerTest.java
+++ b/community/csv/src/test/java/org/neo4j/csv/reader/BufferedCharSeekerTest.java
@@ -36,6 +36,7 @@ import java.util.List;
 import java.util.Random;
 
 import static java.lang.String.format;
+import static java.lang.System.currentTimeMillis;
 import static java.util.Arrays.asList;
 import static org.junit.Assert.assertArrayEquals;
 import static org.junit.Assert.assertEquals;
@@ -49,6 +50,25 @@ import static org.neo4j.helpers.collection.Iterators.array;
 @RunWith( Parameterized.class )
 public class BufferedCharSeekerTest
 {
+    private static final char[] WHITESPACE_CHARS = {
+            Character.SPACE_SEPARATOR,
+            Character.PARAGRAPH_SEPARATOR,
+            '\u00A0',
+            '\u2007',
+            '\u202F',
+            '\t',
+            '\f',
+            '\u001C',
+            '\u001D',
+            '\u001E',
+            '\u001F'
+    };
+
+    private static final char[] DELIMITER_CHARS = {
+            ',',
+            '\t'
+    };
+
     private static final String TEST_SOURCE = "TestSource";
     private final boolean useThreadAhead;
 
@@ -684,6 +704,7 @@ public class BufferedCharSeekerTest
         int columns = random.nextInt( 10 ) + 5;
         int lines = 100;
         List<String> expected = new ArrayList<>();
+        char delimiter = randomDelimiter();
         for ( int i = 0; i < lines; i++ )
         {
             for ( int j = 0; j < columns; j++ )
@@ -693,13 +714,13 @@ public class BufferedCharSeekerTest
                     if ( random.nextBoolean() )
                     {
                         // Space before delimiter
-                        builder.append( " " );
+                        builder.append( randomWhitespace( delimiter ) );
                     }
-                    builder.append( "," );
+                    builder.append( delimiter );
                     if ( random.nextBoolean() )
                     {
                         // Space before delimiter
-                        builder.append( " " );
+                        builder.append( randomWhitespace( delimiter ) );
                     }
                 }
                 boolean quote = random.nextBoolean();
@@ -712,7 +733,7 @@ public class BufferedCharSeekerTest
                         if ( random.nextBoolean() )
                         {
                             // Space after quote start
-                            value += " ";
+                            value += randomWhitespace( delimiter );
                         }
                     }
                     // Actual value
@@ -722,7 +743,7 @@ public class BufferedCharSeekerTest
                         if ( random.nextBoolean() )
                         {
                             // Space before quote end
-                            value += " ";
+                            value += randomWhitespace( delimiter );
                         }
                     }
                     expected.add( value );
@@ -748,15 +769,31 @@ public class BufferedCharSeekerTest
                 String nextExpected = next.next();
                 if ( nextExpected == null )
                 {
-                    assertNextValueNotExtracted( seeker, mark, COMMA );
+                    assertNextValueNotExtracted( seeker, mark, delimiter );
                 }
                 else
                 {
-                    assertNextValue( seeker, mark, COMMA, nextExpected );
+                    assertNextValue( seeker, mark, delimiter, nextExpected );
                 }
             }
         }
-        assertEnd( seeker, mark, COMMA );
+        assertEnd( seeker, mark, delimiter );
+    }
+
+    private char randomDelimiter()
+    {
+        return DELIMITER_CHARS[random.nextInt( DELIMITER_CHARS.length )];
+    }
+
+    private char randomWhitespace( char except )
+    {
+        char ch;
+        do
+        {
+            ch = WHITESPACE_CHARS[random.nextInt( WHITESPACE_CHARS.length )];
+        }
+        while ( ch == except );
+        return ch;
     }
 
     @Test

--- a/community/csv/src/test/java/org/neo4j/csv/reader/BufferedCharSeekerTest.java
+++ b/community/csv/src/test/java/org/neo4j/csv/reader/BufferedCharSeekerTest.java
@@ -19,6 +19,7 @@
  */
 package org.neo4j.csv.reader;
 
+import org.apache.commons.lang3.StringUtils;
 import org.junit.After;
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -30,6 +31,7 @@ import java.io.Reader;
 import java.io.StringReader;
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Random;
 
@@ -42,6 +44,7 @@ import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
 import static org.neo4j.csv.reader.CharSeekers.charSeeker;
 import static org.neo4j.csv.reader.Readables.wrap;
+import static org.neo4j.helpers.collection.Iterators.array;
 
 @RunWith( Parameterized.class )
 public class BufferedCharSeekerTest
@@ -272,10 +275,7 @@ public class BufferedCharSeekerTest
         assertNextValue( seeker, mark, COMMA, "dos" );
         assertNextValue( seeker, mark, COMMA, "tres" );
         assertNextValueNotExtracted( seeker, mark, COMMA );
-        assertTrue( mark.isEndOfLine() );
-
-        // THEN
-        assertFalse( seeker.seek( mark, COMMA ) );
+        assertEnd( seeker, mark, COMMA );
     }
 
     @Test
@@ -436,8 +436,7 @@ public class BufferedCharSeekerTest
         assertNextValue( seeker, mark, COMMA, "four" );
         assertNextValue( seeker, mark, COMMA, "five" );
         assertNextValue( seeker, mark, COMMA, "s\"ix" );
-        assertTrue( mark.isEndOfLine() );
-        assertFalse( seeker.seek( mark, COMMA ) );
+        assertEnd( seeker, mark, COMMA );
     }
 
     @Test
@@ -457,8 +456,7 @@ public class BufferedCharSeekerTest
         assertNextValue( seeker, mark, COMMA, "abc" );
         assertNextValue( seeker, mark, COMMA, "def" );
         assertNextValue( seeker, mark, COMMA, "ghi" );
-        assertTrue( mark.isEndOfLine() );
-        assertFalse( seeker.seek( mark, COMMA ) );
+        assertEnd( seeker, mark, COMMA );
     }
 
     @Test
@@ -526,8 +524,7 @@ public class BufferedCharSeekerTest
         assertNextValue( seeker, mark, COMMA, "4" );
         assertNextValue( seeker, mark, COMMA, "The Cardigans" );
         assertNextValue( seeker, mark, COMMA, "1992" );
-        assertTrue( mark.isEndOfLine() );
-        assertFalse( seeker.seek( mark, COMMA ) );
+        assertEnd( seeker, mark, COMMA );
     }
 
     @Test
@@ -663,6 +660,141 @@ public class BufferedCharSeekerTest
         assertNextValue( seeker, mark, COMMA, "Waaaa" );
     }
 
+    @Test
+    public void shouldTrimStringsWithFirstLineCharacterSpace() throws IOException
+    {
+        // given
+        String line = " ,a, ,b, ";
+        seeker = seeker( line, withTrimStrings( config(), true ) );
+
+        // when/then
+        assertNextValueNotExtracted( seeker, mark, COMMA );
+        assertNextValue( seeker, mark, COMMA, "a" );
+        assertNextValueNotExtracted( seeker, mark, COMMA );
+        assertNextValue( seeker, mark, COMMA, "b" );
+        assertNextValueNotExtracted( seeker, mark, COMMA );
+        assertEnd( seeker, mark, COMMA );
+    }
+
+    @Test
+    public void shouldParseAndTrimRandomStrings() throws IOException
+    {
+        // given
+        StringBuilder builder = new StringBuilder();
+        int columns = random.nextInt( 10 ) + 5;
+        int lines = 100;
+        List<String> expected = new ArrayList<>();
+        for ( int i = 0; i < lines; i++ )
+        {
+            for ( int j = 0; j < columns; j++ )
+            {
+                if ( j > 0 )
+                {
+                    if ( random.nextBoolean() )
+                    {
+                        // Space before delimiter
+                        builder.append( " " );
+                    }
+                    builder.append( "," );
+                    if ( random.nextBoolean() )
+                    {
+                        // Space before delimiter
+                        builder.append( " " );
+                    }
+                }
+                boolean quote = random.nextBoolean();
+                if ( random.nextBoolean() )
+                {
+                    String value = "";
+                    if ( quote )
+                    {
+                        // Quote
+                        if ( random.nextBoolean() )
+                        {
+                            // Space after quote start
+                            value += " ";
+                        }
+                    }
+                    // Actual value
+                    value += String.valueOf( random.nextInt() );
+                    if ( quote )
+                    {
+                        if ( random.nextBoolean() )
+                        {
+                            // Space before quote end
+                            value += " ";
+                        }
+                    }
+                    expected.add( value );
+                    builder.append( quote ? "\"" + value + "\"" : value );
+                }
+                else
+                {
+                    expected.add( null );
+                }
+            }
+            builder.append( format( "%n" ) );
+        }
+        String data = builder.toString();
+        seeker = seeker( data, withTrimStrings( config(), true ) );
+
+        // when
+        Iterator<String> next = expected.iterator();
+        for ( int i = 0; i < lines; i++ )
+        {
+            for ( int j = 0; j < columns; j++ )
+            {
+                // then
+                String nextExpected = next.next();
+                if ( nextExpected == null )
+                {
+                    assertNextValueNotExtracted( seeker, mark, COMMA );
+                }
+                else
+                {
+                    assertNextValue( seeker, mark, COMMA, nextExpected );
+                }
+            }
+        }
+        assertEnd( seeker, mark, COMMA );
+    }
+
+    @Test
+    public void shouldParseNonLatinCharacters() throws IOException
+    {
+        // given
+        List<String[]> expected = asList(
+                array( "普通�?/普通話", "\uD83D\uDE21" ),
+                array( "\uD83D\uDE21\uD83D\uDCA9\uD83D\uDC7B", "ⲹ楡�?톜ഷۢ⼈�?�늉�?�₭샺ጚ砧攡跿家䯶�?⬖�?�犽ۼ" ),
+                array( " 㺂�?鋦毠", ";먵�?裬岰鷲趫\uA8C5얱㓙髿ᚳᬼ≩�?� " )
+        );
+        String data = lines( format( "%n" ), expected );
+
+        // when
+        seeker = seeker( data );
+
+        // then
+        for ( String[] line : expected )
+        {
+            for ( String cell : line )
+            {
+                assertNextValue( seeker, mark, COMMA, cell );
+            }
+        }
+        assertEnd( seeker, mark, COMMA );
+    }
+
+    private String lines( String newline, List<String[]> cells )
+    {
+        String[] lines = new String[cells.size()];
+        int i = 0;
+        for ( String[] columns : cells )
+        {
+            lines[i++] = StringUtils.join( columns, "," );
+        }
+        return lines( newline, lines );
+    }
+
     private String lines( String newline, String... lines )
     {
         StringBuilder builder = new StringBuilder();
@@ -751,6 +883,12 @@ public class BufferedCharSeekerTest
     {
         assertTrue( seeker.seek( mark, delimiter ) );
         assertFalse( seeker.tryExtract( mark, extractors.string() ) );
+    }
+
+    private void assertEnd( CharSeeker seeker, Mark mark, int delimiter ) throws IOException
+    {
+        assertTrue( mark.isEndOfLine() );
+        assertFalse( seeker.seek( mark, delimiter ) );
     }
 
     private String[] nextLineOfAllStrings( CharSeeker seeker, Mark mark ) throws IOException

